# Basic data samples 

Making data samples using the Sample module.



## Atoms and pure data 

First, let's sample data consisting of just sequences of (:) atoms and then samples of all pure data with 

In [1]:
from base import *
import Sample 

In [2]:
#
#   Use oddAtoms and evenAtoms to get datas consisting of sequences of (:). 
for d in Sample. oddAtoms(4): print(len(d),d)
for d in Sample.evenAtoms(4): print(len(d),d)

1 ◎
3 ◎ ◎ ◎
5 ◎ ◎ ◎ ◎ ◎
7 ◎ ◎ ◎ ◎ ◎ ◎ ◎
0 
2 ◎ ◎
4 ◎ ◎ ◎ ◎
6 ◎ ◎ ◎ ◎ ◎ ◎


In [3]:
#
#   For pure samples, use pure(width,depth) for all pure data with width 
#   less than or equal to 'width' and depth less than or equal to 'depth'. 
#
import Set 
def pure(width,depth):
    L = Sample.pure(width,depth)
    return Set.Set(*L)
print(pure(5,2))

62193781


### Pure Data Sizes 

|         | width 1  | width 2 | width 3 | width 4  | width 5 |
|---------|---------:|--------:|--------:|---------:|--------:|
|depth 1  |    2     |    3    |    4    |     5    |    6    |
|depth 2  |    5     |     91   |    4369   |   406,901    |    62,193,781    |
|depth 3  |    26     |     68,583,243   |    ?    |     ?    |    ?    |
|depth 4  |    677     |     ?   |    ?    |     ?    |    ?    |
|depth 5  |    458,330     |     ?   |    ?    |     ?    |    ?    |

The sizes of these samples grows very rapidly with maximum width and depth. 

In [4]:
for s in pure(3,2).sample(10): print(s)

(:◎ ◎) (:◎) ◎
◎◎ ◎ 
  𝟬
◎◎◎ (:◎ ◎) ◎
◎◎ ◎◎ (:◎ ◎)
◎◎◎ ◎◎ 
◎◎ ◎ ◎◎◎
◎ ◎ ◎
◎ ◎◎ 𝟬
◎◎◎ 𝟬


## Coda based samples 

The Domain class in Sample lets you get data samples using a given collections of codas.  You specify a sample with a list of codas, and a specified maximum width and maximum depth.

In [5]:
#
#   For example, let's make data samples 
#   with 'a' and 'b' for base codas. 
#
DOM = Sample.Domain(co('a'),co('b'))
def window(D,width,depth):
    W = Set.Set(*D.window(width,depth))
    return W
for s in window(DOM,2,2): print(s)


a
b
◎
(:a)
(:b)
(a:)
(a:a)
(a:b)
(b:)
(b:a)
(b:b)
a
a a
a b
a ◎
a (:a)
a (:b)
a (a:)
a (a:a)
a (a:b)
a (b:)
a (b:a)
a (b:b)
b
b a
b b
b ◎
b (:a)
b (:b)
b (a:)
b (a:a)
b (a:b)
b (b:)
b (b:a)
b (b:b)
◎
◎ a
◎ b
◎ ◎
◎ (:a)
◎ (:b)
◎ (a:)
◎ (a:a)
◎ (a:b)
◎ (b:)
◎ (b:a)
◎ (b:b)
(:a)
(:a) a
(:a) b
(:a) ◎
(:a) (:a)
(:a) (:b)
(:a) (a:)
(:a) (a:a)
(:a) (a:b)
(:a) (b:)
(:a) (b:a)
(:a) (b:b)
(:b)
(:b) a
(:b) b
(:b) ◎
(:b) (:a)
(:b) (:b)
(:b) (a:)
(:b) (a:a)
(:b) (a:b)
(:b) (b:)
(:b) (b:a)
(:b) (b:b)
(a:)
(a:) a
(a:) b
(a:) ◎
(a:) (:a)
(a:) (:b)
(a:) (a:)
(a:) (a:a)
(a:) (a:b)
(a:) (b:)
(a:) (b:a)
(a:) (b:b)
(a:a)
(a:a) a
(a:a) b
(a:a) ◎
(a:a) (:a)
(a:a) (:b)
(a:a) (a:)
(a:a) (a:a)
(a:a) (a:b)
(a:a) (b:)
(a:a) (b:a)
(a:a) (b:b)
(a:b)
(a:b) a
(a:b) b
(a:b) ◎
(a:b) (:a)
(a:b) (:b)
(a:b) (a:)
(a:b) (a:a)
(a:b) (a:b)
(a:b) (b:)
(a:b) (b:a)
(a:b) (b:b)
(b:)
(b:) a
(b:) b
(b:) ◎
(b:) (:a)
(b:) (:b)
(b:) (a:)
(b:) (a:a)
(b:) (a:b)
(b:) (b:)
(b:) (b:a)
(b:) (b:b)
(b:a)
(b:a) a
(b:a) b
(b:a) ◎
(b:a) (:a)
(b

In [6]:
#
#   For a more interesting sample, 
#
DOM2 = Sample.Domain(*Sample.codas)
print(len(Sample.codas),','.join([str(c) for c in Sample.codas]))

62 ◎,𝟬,𝝞,nat,one,count,int1,nat1,int_add,int_mult,int_diff,int_max,int_min,int_sort,int_inv,pass,null,bin,domain,if,put,has1,get1,*,some,rev,first,tail,last,rep,nth1,pre,post,bool,not,=,or,and,nor,xor,nand,xnor,iff,imply,ap,app,ap2,aps,apby,apif,equiv,collect,counts,0,1,2,3,a,b,A,B,?


In [7]:
S22 = window(DOM2,2,2)

In [8]:
for s in S22.sample(10): print(s)

(*:nat) (and:nat1)
(int1:aps) (domain:domain)
(post:or) app
(nand:imply) (iff:=)
(put:count) one
(a:?) (=:nat)
(tail:domain) (put:post)
(0:or) (2:pass)
(aps:int_inv) (or:nat)
(tail:int_inv) (apif:pre)


In [9]:
for s in window(DOM2,3,1).sample(10): print(s)

some ?
collect xnor 𝝞
ap a one
and iff or
aps app nat
or collect int_inv
some iff =
nat1 get1 apby
ap2 3 xor
A post bool


In [10]:
def pr(w,d): print(w,d,len(window(DOM2,w,d)))

In [16]:
pr(1,3)

1 3 16261056


### Coda Sample Sizes  

|         | width 1  | width 2 | width 3 | width 4  | width 5 |
|---------|---------:|--------:|--------:|---------:|--------:|
|depth 1  |    63     |    3969    |    250,047    |     15,752,961    |    ?    |
|depth 2  |    4032    |    16,257,024   |     ?   |      ?    |    ?    |
|depth 3  |    16,261,056     |     ?   |    ?    |     ?    |    ?    |
|depth 4  |    ?     |     ?   |    ?    |     ?    |    ?    |
|depth 5  |    ?     |     ?   |    ?    |     ?    |    ?    |

In either case of pure data or coda samples like the above, any data will eventually appear in the sample for large enough width and depth.  Coda samples grow rapidly with width and depth as well, but they are far more practical for 
searching since the base set of codas can be used to aim towards known definitions. 